bioimageio/core/build_spec/build_spec.py

```diff
@@ -0,0 +1,440 @@
+        # for floating point numbers we assume valid range from -inf to inf
+        elif np.issubdtype(np.dtype(dtype), np.floating):
+            min_, max_ = -np.inf, np.inf
+        elif np.issubdtype(np.dtype(dtype), np.bool):
+            min_, max_ = 0, 1
+        else:
+            raise RuntimeError(f"Cannot derived data range for dtype {dtype}")
+        data_range = (min_, max_)
+    assert isinstance(data_range, (tuple, list))
+    assert len(data_range) == 2
+    return data_range
+
+
+def _get_axes(axes, ndim):
+    if axes is None:
+        assert ndim in (2, 4, 5)
+        default_axes = {2: "bc", 4: "bcyx", 5: "bczyx"}
+        axes = default_axes[ndim]
+    return axes
+
+
+def _get_input_tensor(test_in, name, step, min_shape, data_range, axes, preprocessing):
+    shape = test_in.shape
+    if step is None:
+        assert min_shape is None
+        shape_description = shape
+    else:
+        shape_description = {"min": shape if min_shape is None else min_shape, "step": step}
+
+    axes = _get_axes(axes, test_in.ndim)
+    data_range = _get_data_range(data_range, test_in.dtype)
+
+    kwargs = {}
+    if preprocessing is not None:
+        kwargs["preprocessing"] = preprocessing
+
+    inputs = model_spec.raw_nodes.InputTensor(
+        name="input" if name is None else name,
+        data_type=str(test_in.dtype),
+        axes=axes,
+        shape=shape_description,
+        data_range=data_range,
+        **kwargs,
+    )
+    return inputs
+
+
+def _get_output_tensor(test_out, name, reference_input, scale, offset, axes, data_range, postprocessing, halo):
+    shape = test_out.shape
+    if reference_input is None:
+        assert scale is None
+        assert offset is None
+        shape_description = shape
+    else:
+        assert scale is not None
+        assert offset is not None
+        shape_description = {"reference_input": reference_input, "scale": scale, "offset": offset}
+
+    axes = _get_axes(axes, test_out.ndim)
+    data_range = _get_data_range(data_range, test_out.dtype)
+
+    kwargs = {}
+    if postprocessing is not None:
+        kwargs["postprocessing"] = postprocessing
+    if halo is not None:
+        kwargs["halo"] = halo
+
+    outputs = model_spec.raw_nodes.OutputTensor(
+        name="output" if name is None else name,
+        data_type=str(test_out.dtype),
+        axes=axes,
+        data_range=data_range,
+        shape=shape_description,
+        **kwargs,
+    )
+    return outputs
+
+
+def _build_authors(authors: List[Dict[str, str]]):
+    return [model_spec.raw_nodes.Author(**a) for a in authors]
+
+
+# TODO The citation entry should be improved so that we can properly derive doi vs. url
+def _build_cite(cite: Dict[str, str]):
+    citation_list = [model_spec.raw_nodes.CiteEntry(text=k, url=v) for k, v in cite.items()]
+    return citation_list
+
+
+# TODO we should make the name more specific: "build_model_spec"?
+# TODO maybe "build_raw_model" as it return raw_nodes.Model
```

Comment:
I think 'build_raw_model' is a good idea, as we might add `build_model` in the future (for now that does not make so much sense as there is no serialization from non-raw nodes...

Reply:
I decided to go with `build_model` here. I find the `raw` things very confusing in the function names.